### ISSUE TITLE
BUG: Direct renaming of 1 column seems to be accepted, but only old name is working

### ISSUE THREAD

**Issue by MarianD** [pandas-dev/pandas#38547]
- [ ] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```python
In[5]: df = pd.DataFrame(np.eye(2), columns=["a", "b"])
In[6]: df
Out[6]: 
     a    b
0  1.0  0.0
1  0.0  1.0

In[7]: df.columns.values[0] = "x"            # <---- HERE! Direct renaming "a" -> "x"
In[8]: df
Out[8]: 
     x    b                                  # Seems renamed (x instead of a)
0  1.0  0.0
1  0.0  1.0

In[9]: df.columns
Out[9]: Index(['x', 'b'], dtype='object')    # Seems renamed ('x' instead of 'a')

In[10]: df["x"]       # KeyError: 'x'        # New name doesn't work

In[11]: df["a"]       # OK                   # Old name indeed still exists
Out[11]: 
0    1.0
1    0.0
Name: a, dtype: float64

```

#### Problem description

After direct renaming a column of a dataframe, all its properties *show* the new name.   
But all attempts to *use* this new name fail with the `KeyError`, while the old, “hidden” name is still functional.

#### Expected Output

Raising an exception for such a direct renaming, or fully accept it as a new name (forgetting about the old one).

#### Output of ``pd.show_versions()``

<details>

    INSTALLED VERSIONS
    ------------------
    commit           : None
    python           : 3.8.3.final.0
    python-bits      : 64
    OS               : Windows
    OS-release       : 7
    machine          : AMD64
    processor        : Intel64 Family 6 Model 42 Stepping 7, GenuineIntel
    byteorder        : little
    LC_ALL           : None
    LANG             : None
    LOCALE           : Slovak_Slovakia.1250

    pandas           : 1.0.5
    numpy            : 1.18.5
    pytz             : 2020.1
    dateutil         : 2.8.1
    pip              : 20.1.1
    setuptools       : 49.2.0.post20200714
    Cython           : 0.29.21
    pytest           : 5.4.3
    hypothesis       : None
    sphinx           : 3.1.2
    blosc            : None
    feather          : None
    xlsxwriter       : 1.2.9
    lxml.etree       : 4.5.2
    html5lib         : 1.1
    pymysql          : None
    psycopg2         : None
    jinja2           : 2.11.2
    IPython          : 7.16.1
    pandas_datareader: None
    bs4              : 4.9.1
    bottleneck       : 1.3.2
    fastparquet      : None
    gcsfs            : None
    lxml.etree       : 4.5.2
    matplotlib       : 3.2.2
    numexpr          : 2.7.1
    odfpy            : None
    openpyxl         : 3.0.4
    pandas_gbq       : None
    pyarrow          : None
    pytables         : None
    pytest           : 5.4.3
    pyxlsb           : None
    s3fs             : None
    scipy            : 1.5.0
    sqlalchemy       : 1.3.18
    tables           : 3.6.1
    tabulate         : None
    xarray           : None
    xlrd             : 1.2.0
    xlwt             : 1.3.0
    xlsxwriter       : 1.2.9
    numba            : 0.50.1

</details>

**Comment by aflah02** [comment]
I was wondering can you tell me where are these functions defined?
I have been unable to navigate through the database
Maybe I can try to fix this

**Comment by MarianD** [comment]
@aflah02 
> 
> I was wondering can you tell me where are these functions defined?
> I have been unable to navigate through the database
> Maybe I can try to fix this

I don't know.

**Comment by aflah02** [comment]
Oh okay thanks for replying

**Comment by mzeitlin11** [comment]
> I was wondering can you tell me where are these functions defined?
> I have been unable to navigate through the database
> Maybe I can try to fix this

Hi @aflah02! Navigating around a large codebase like this is difficult - I'd recommend learning how to use a debugger (like https://docs.python.org/3/library/pdb.html or something more convenient like the built-in debugger provided by PyCharm). This allows you to step through what happens when code is run step by step, which will show where the relevant code lives.

**Comment by jreback** [comment]
I am not sure why you think this would work. Indexes are immutable objects. Sure you *can* do things like this and they *may* work, but no-where in the documentation would you in any way expect this to work.

**Comment by simonjayhawkins** [comment]
yeah, so the issue is that `df.columns.values` should be a copy and not a view so that the parent DataFrame/immutable Index is not updated?

**Comment by MarianD** [comment]
@simonjayhawkins, I think it would be a nice solution supposing it would be documented and not breaking the frequently used paradigm `df.columns = iterable_of_ALL_column_names`.

**Comment by jreback** [comment]
we should simply return a read-only numpy array here to follow semantics. -1 on actually allowing a view on an immutable object that you can change.
